### ISSUE TITLE
Steps for bringing the newest build to OSX

### ISSUE THREAD

**Issue by aktau** [PCSX2/pcsx2#18]
Hey  guys, this is just some exploratory thinking I'm doing but:

Seeing as PCSX2 supports both Linux and Windows and is (as far as I can see) based mostly on cross-platform libs. This leads to the suspicion that a proper port to OSX wouldn't be a herculanean task. The (free) OSX 10.9 supports OpenGL 4.1 and the drivers are "ok" (not great). Given the fact that 10.9 is free and the OpenGL implementation is far less buggy than earlier incarnations, I see no reason to support anything earlier.

Instead of plowing through the codebase, I'd like to ask the devs: what do you see as the major hurdles that someone porting PCSX2 would encounter?

**EDIT**: I'm aware of earlier versions of PCSX2 being made available for OSX and I've noticed some vestigial bits in the CMake files et al.

**Comment by juhalaukkanen** [comment]
Just to note that this is still alive. I don't have too much time per week to work with this so progress is rather slow.

https://github.com/juhalaukkanen/pcsx2/commits/master

**Comment by avih** [comment]
I'm copying my question from #304 : 

Care to give a some update on the current status?
- On what system do you compile and which XCode version?
- Any special flags/modifications/etc instructions to get it to build?
- What builds successfully? PCSX2 itself? plugins? which?
- Does it run? what works and what doesn't?
- Any other info? does it run in reasonable speed? any gotchas?
- Do you have patches you're happy with and can do a PR?

**Comment by juhalaukkanen** [comment]
This is a merge between my work and aktau's.

- I have compiled against Darwin 13.x (Mavericks) (Macbook Pro native) and 14.x (Yosemite) (VirtualBox VM). Xcode 5.x/6x SDK required since with fresh c++0x / stdlib features
- Some Xcode versions seem to require explicit -msse -msse2 -msse3 switches
- PCSX2 builds and all null plugins.
-Some runtime errors remain and GSdx/onepad are not yet ported. Also none of the GTK dialogs do not yet have WX equivalent implementation for OSX.
-also really big issue with dynarec mmap so only interpreter core works now (seems to be 32bit Darwin mmap issue)
-spu2x has some minor related issues like homebrew not providing 32bit sound-touch build and I have not yet tested it well

**Comment by avih** [comment]
Cheers, good info. I'd imagine the trickiest part would be to make the recompiler work.

As for the gtk dialogs, TBH they're just dialogs to end up with a gsdx.ini file. One could as well make the config a stand alone application which gsdx invokes and then waits for it to complete, then reads the ini file.

**Comment by uyjulian** [comment]
@juhalaukkanen No 32 bit/universal option for libs on homebrew? Here, send a pull request to the homebrew repo (https://github.com/Homebrew/homebrew)

**Comment by aktau** [comment]
> On what system do you compile and which XCode version?
> Any special flags/modifications/etc instructions to get it to build?
> What builds successfully? PCSX2 itself? plugins?
> Does it run? what works and what doesn't?
> Any other info? does it run in reasonable speed? any gotchas?
> Do you have patches you're happy with and can do a PR?

First of all, apologies for my long silence. I have a new job and I've moved countries. Things are still not as quiet as they should be and I've still got much to learn at work so I don't expect to have loads of free time.

As indicated by @juhalaukkanen we started working on porting PCSX2 to OSX in parallel, I guess sometimes taking hints from each other (I know I peeked at his branch a few times :). Our philosophy was/is a bit different though:

- OSX tooling for 32-bit is a bit weak, and also an unknown for me (homebrew, the unofficial OSX package manager doesn't "do" 32-bit, for example). 
- OSX (Apple) has been trying to get rid of X11 for a decent time, and they're making it more difficult to install. That, and a "non-native" solution feels unclean to me.

So I started my porting work by wholesale porting the codebase to work on 64-bit OSX (PR #297, in the top post you can see what's "done", e.g.: some libraries compile). Needless to say, this means I'm unable to test anything at all until I get a binary compiled. And most likely when that happens, the binary will instantly crash.

A few courses of actions that make more sense:

1) Get 64-bit working on Linux, then port to OSX (this is advocated by @gregory38)
2) Get PCSX2 32-bit working on OSX, then get it working (this is being done by @juhalaukkanen)

To counteract the fact that my approach is the crazier one, I'm trying to create small, focused commits so they're easily reviewable. That doesn't really offset anything of course.

**Comment by juhalaukkanen** [comment]
Here is now latest status: https://github.com/juhalaukkanen/pcsx2/commits/master

- SPU2x ported but not tested well. I have submitted pull req to homebrew for 32bit sound-touch
- GSdx has now SDL2 OpenGL backend which can be utilized on OSX/Linux. Possibly Windows also. I have not tested this with Linux. OSX is unsuable because of mmap issue (see below)

I have not been able to solve mmap problem in 32bit OSX. It causes random runtime pagefaults and GSdx cannot even execute properly. Any help on this topic really much appreciated.

**Comment by rodneyrod** [comment]
How's the progress on this going?

**Comment by uyjulian** [comment]
I don't think the repo updated since February.

Little bits:
It's fine to port PCSX2 to 32-bit OSX, but OpenGL is only up to 4.1.
I doubt that Apple will update OpenGL anymore because now they are pushing Metal on Macs now.
GSdx still requires OpenGL 4.5 even for software mode.

**Comment by juhalaukkanen** [comment]
Progress slowly. Now I have been trying to hunt down those crashes. Finding the root is really hard. I can avoid one crash by adding sleep delays in appropriate areas so it is a race condition between threads which is always time consuming. The other crash is with recompiler not yet working properly.

So is this OpenGL 4.5 really hard requirement? I see GLES support there also and it should be able work without that high context?

**Comment by gregory38** [comment]
The requirement is gl 3.3 + a couple (2) gl 4.2 extensions. All others gl4 extensions are a bonus to improve speed/accuracy.
I really hope apple will implement newer gl extensions. At least some of them.

**Comment by aktau** [comment]
They already introduced OSX 10.11 (El Capitan) and did not mention OpenGL afaik. Historically, this indicates they won't be bumping the OpenGL so @uyjulian is right. There is a very faint possibility that there are more OpenGL extensions on 10.11, but I haven't seen anything like that mentioned online, sadly.

**Comment by shiggitay** [comment]
What's the status of this whole thing? I'd be willing to help test any builds you guys come up with to help keep and get this going... I've been trying to get the latest 1.3.1 Git build working under WineSkin without much luck. Just so this is out there: I'm not developer... I'm only a casual gamer wanting to better the PS2 emulation scene for all. Thanks.

**Comment by uyjulian** [comment]
GL_ARB_shading_language_420pack isn't on Mac OS 10.11.
However GL_ARB_texture_storage is in Mac OS 10.11.

There's really no update because porting PCSX2 to 64-bit would take a lot of work.
The recompiler is really ugly too.

@shiggitay check http://forums.pcsx2.net/Thread-Testers-Wanted-PCSX2-CE-for-OS-X-v1-0?action=newpost for latest information about making PCSX2 working under wineskin

**Comment by shiggitay** [comment]
@uyjulian Yeah I know of that project, That user posts on a site I used to frequent (EmuScene, though nowadays it's called MacScene) and he's got a similar thread there as he does on the main PCSX2 site. I'll read into it more when I have some time. Thanks for the response and pseudo-update. lol :)

**Comment by gregory38** [comment]
Honestly it sucks. It would cost nothing to support recent openGL. They just do it to locked-in their clients.

@juhalaukkanen 
```
I have not been able to solve mmap problem in 32bit OSX. It causes random runtime pagefaults and GSdx cannot even execute properly. Any help on this topic really much appreciated.
```
Are you sure the runtime pagefaults aren't expected. PCSX2 uses pagefaults to handle buffer allocation and to detect self-modyfied code. The former will less appear in the future (see #967).

**Comment by juhalaukkanen** [comment]
Might be that..

Anyway I am thinking of pushing this in current state so it compiles but runtime errors. Is this OK? Otherwise code just diverges too much..

**Comment by gregory38** [comment]
Linux port wasn't done in 1 day. I'm in favor to merge as much code as possible. However we want to do a release soon. So I propose to you that you do a pr. We select all the safe commits and merge them. The other will be merged later. It will reduce the maintenance burden on your side.

**Comment by gregory38** [comment]
FWIW I merged my branch. The number of expected SIGSEGV is greatly reduce. It mostly remains the one to protect the EE memory (address 0x2000_00000, with a 32MB area).

**Comment by alexchandel** [comment]
I wouldn't ever expect 32-bit PCSX2 to work perfectly on OSX. The i686 api is second-class to the x86_64, and Apple discourages developing new software against it. I wouldn't be surprised if it has anomalies. FWIW Dolphin/Mupen64Plus/Mednafen/Higan/DeSmuME/OpenEmu/everyone except Project 64 (& who cares, Mupen64Plus replaced them) targets x86_64 OS X, and I believe some never explicitly targeted i686 at all.

**Comment by gregory38** [comment]
Seriously, could people understand that a complex port must be done step by step. None of intermediate step will be the final port. So we don't care about others emulator. Neither we care about the current situation of any libs/arch on OSx.

As a side note, I think that recompiler's issues are linked with the OSx abi. I have some code to factorize stack frame creation so it would be easier to fix all recompilers in one shot.

**Comment by shiggitay** [comment]
Has any progress been made on this since late 2015? I tried to compile @juhalaukkanen's fork but it died on what I think was the GSDX plugin. I'm running macOS 10.13 High Sierra GM Candidate with XCode 9 beta. I'm still willing to help you guys get this working with modern code on macOS (as a build tester ofc, since I can't code for beans). I'd like to record some gameplay for a YouTube channel I have, but running v1.4.0 or 1.5.0 git under WINE makes it run all glitchy, or not well at all.

BTW, here's what happens when I try and compile that forked code:

https://bpaste.net/show/2315e29a1bc7

**Comment by willkuer** [comment]
Even if you are abke to compile it that doesn't mean you can run it. Even if you can compile and run it that doesn't mean that you can run it with the target framerate.

First pcsx2 needs to be ported to x64 which is IIRC already done or rather easy to achieve for the interpreter and gsdx software mode. The recompilers however seem to be much harder.

When the recompiler and gsdx hw mode are ported to x64 one can consider to port it to mac and I guess it would be rather easy.

x64 port is currently low priority.

**Comment by shiggitay** [comment]
@willkuer oh I see. Well I've not given up hope. I guess I'll wait it out and see what happens.

**Comment by uyjulian** [comment]
OpenGL is deprecated but is not taken out of macOS yet.

Future progress on Apple platforms involves 64-bit JIT (which I doubt will get finished soon) and Metal (which can be solved with MoltenVK if a Vulkan renderer get implemented).

Good news: Linux runs on the newest Macs with T2 chip just fine; you just need to use an external HDD since the T2 chip does not allow internal disk access on unsigned OS.

**Comment by bryce13950** [comment]
Now that the next version of macOS is dropping 32 bit apps, it seems pretty apparent that this is no longer going to be able to be ran on macOS at all even in Wine. Is there any progress on a macOS port? Is there anything that I could do to help make it possible?

**Comment by tadanokojin** [comment]
No progress is being made for 1.6.0

You can try getting the build working on MacOS. That would be a step.

**Comment by bryce13950** [comment]
> No progress is being made for 1.6.0
> 
> You can try getting the build working on MacOS. That would be a step.

Are there instructions anywhere on getting the project setup for linux development?

**Comment by gregory38** [comment]
The not funny part. Apple 64bits != Linux 64bits != Windows 64bits !!!

**Comment by uyjulian** [comment]
You can build the project on Linux using the usual cmake commands.

macOS won't happen anytime soon. Well, unless you can port the Linux-specific functions to Mach-specific functions. Even if you did that, the recompiler is still 32-bit, and a 64-bit port of the recompiler won't be happening for a while.

**Comment by arcum42** [comment]
On Linux, what I would usually do to compile pcsx2 as 64 bit would be:
sh build.sh --clean --no-cross-multilib
Lately, I've been adding --clang to the arguments as well. However, I would then go into PCSX2, emulation settings, uncheck Preset, under EE/IOP, set both EmotionEngine and IOP to Interpreter, and under VUs, set VU0 and VU1 to Interpreter.

Having any of those settings on a recompiler will crash a 64 bit build. On a Mac, I imagine getting it to compile that far, even, will be a challenge, though.

**Comment by bryce13950** [comment]
> You can build the project on Linux using the usual cmake commands.
> 
> macOS won't happen anytime soon. Well, unless you can port the Linux-specific functions to Mach-specific functions. Even if you did that, the recompiler is still 32-bit, and a 64-bit port of the recompiler won't be happening for a while.

Is there anybody currently working on the 64-bit port? What needs to be done? I am all in to help make it possible.

**Comment by bryce13950** [comment]
> On Linux, what I would usually do to compile pcsx2 as 64 bit would be:
> sh build.sh --clean --no-cross-multilib
> Lately, I've been adding --clang to the arguments as well. However, I would then go into PCSX2, emulation settings, uncheck Preset, under EE/IOP, set both EmotionEngine and IOP to Interpreter, and under VUs, set VU0 and VU1 to Interpreter.
> 
> Having any of those settings on a recompiler will crash a 64 bit build. On a Mac, I imagine getting it to compile that far, even, will be a challenge, though.

Thanks for the advice. I think I will start by getting it to compile for linux before trying it on macOS

**Comment by uyjulian** [comment]
64-bit interpreter already works. However, getting a 64-bit recompiler to work requires knowledge of AMD64, r3000, and r5900 assembly.

**Comment by tellowkrinkle** [comment]
<img width="640" alt="image" src="https://user-images.githubusercontent.com/3315070/79183264-9b49cd80-7dd6-11ea-8d8a-600bf5da299f.png">

**Comment by shiggitay** [comment]
> Clearly colors are the most important so I bring you... colors
> <img alt="image" width="640" src="https://user-images.githubusercontent.com/3315070/79183264-9b49cd80-7dd6-11ea-8d8a-600bf5da299f.png">
> Current status:
> 
> * 64-bit interpreter working
> * GSdx working
> * onepad not yet working
> * spu2-x not yet working
> * Application freezes if you try to quit it (requires force quit)
> 
> If someone has a version of macOS that still supports 32-bit apps and wants to test to see if the JIT works: https://github.com/tellowkrinkle/pcsx2/tree/MacOS
> At the moment I'm on Catalina so I'll be sticking with my 20% speed for now

You doing the lord's work! Is that a native binary?

**Comment by uyjulian** [comment]
macOS Catalina has functionality for 32-bit code fragments to be run in a 64-bit process (this allows `wine32on64` to work). It may be possible that this functionality could be used to allow the 32 bit JIT to run in a 64 bit process.

**Comment by tellowkrinkle** [comment]
> You are doing the lord's work! Is that a native binary?
> 
> Also I do have a copy of Mojave available to me on my AMD rig here (Hackintosh). I could swap in my AMD GPU to see how it runs.

Yep it's native (hopefully no one minds my usage of ObjC++ for getting an OpenGL context but I can try to do something else later if I need to)

If you're going to try to build a 32-bit binary just be prepared to compile every non-system dependency from source as Homebrew doesn't ship them.  To give an example, here's what the 32-bit wine package's build looked like before Homebrew dropped it https://github.com/Homebrew/homebrew-core/blob/ece7323f93ac7a52fe56133af9e79aecb3bd5530/Formula/wine.rb

(And remember, no audio or input yet so it's not the most useful, though I do plan to work on those next)

A few other things to note
- I bumped the wxWidgets version to 3.1 in a lazy attempt to fix a bug that was causing crashes when attempting to load specific languages in the load-all-languages-and-see-which-ones-work workaround using wxMac 3.0.4.  Annoyingly wx3.1 added additional validation that verifies that you don't try to put right-aligned buttons in horizontally-stretching containers (or bottom-aligned in vertically-stretching) which kind of disagrees with our "apply one style to all buttons" way of doing things.  Queue sketchy workaround.  If that's not preferred I can see if the bug was fixed in later 3.0.x releases instead.
- I had to bring two mandatory OpenGL extensions optional workarounds back from the dead (`GL_ARB_shading_language_420pack` and `GL_ARB_buffer_storage`)
- Whatever version of GDK the GSdx config panel uses looks absolutely disgusting on macOS

And finally a question to the other pcsx2 devs:
What was this put here for?
https://github.com/PCSX2/pcsx2/blob/49e316e488c224bf50df0ee1e3bfbc55fa738343/plugins/GSdx/stdafx.cpp#L158-L160
macOS Catalina didn't like it (any attempts to `mmap` with that flag would fail) and removing it hasn't blown anything up yet, but I feel like that was at least important at some point...

**Comment by shiggitay** [comment]
> > You are doing the lord's work! Is that a native binary?
> > Also I do have a copy of Mojave available to me on my AMD rig here (Hackintosh). I could swap in my AMD GPU to see how it runs.
> 
> Yep it's native (hopefully no one minds my usage of ObjC++ for getting an OpenGL context but I can try to do something else later if I need to)
> 
> If you're going to try to build a 32-bit binary just be prepared to compile every non-system dependency from source as Homebrew doesn't ship them. To give an example, here's what the 32-bit wine package's build looked like before Homebrew dropped it https://github.com/Homebrew/homebrew-core/blob/ece7323f93ac7a52fe56133af9e79aecb3bd5530/Formula/wine.rb
> Edit: Oh yeah, and at the moment you'll need to rename all the plugins from `.so` to `.dylib` because macOS uses a different file extension from linux and otherwise PCSX2 will ignore them. You'll also want to build with `--extra` (or `-DEXTRA_PLUGINS=ON` if you're invoking cmake directly) for the null audio and input plugins
> 
> (And remember, no audio or input yet so it's not the most useful, though I do plan to work on those next)
> 
> A few other things to note
> 
> * I bumped the wxWidgets version to 3.1 in a lazy attempt to fix a bug that was causing crashes when attempting to load specific languages in the load-all-languages-and-see-which-ones-work workaround using wxMac 3.0.4.  Annoyingly wx3.1 added additional validation that verifies that you don't try to put right-aligned buttons in horizontally-stretching containers (or bottom-aligned in vertically-stretching) which kind of disagrees with our "apply one style to all buttons" way of doing things.  Queue sketchy workaround.  If that's not preferred I can see if the bug was fixed in later 3.0.x releases instead.
> * I had to bring two mandatory OpenGL extensions optional workarounds back from the dead (`GL_ARB_shading_language_420pack` and `GL_ARB_buffer_storage`)
> * Whatever version of GDK the GSdx config panel uses looks absolutely disgusting on macOS
> 
> And finally a question to the other pcsx2 devs:
> What was this put here for?
> https://github.com/PCSX2/pcsx2/blob/49e316e488c224bf50df0ee1e3bfbc55fa738343/plugins/GSdx/stdafx.cpp#L158-L160
> 
> 
> macOS Catalina didn't like it (any attempts to `mmap` with that flag would fail) and removing it hasn't blown anything up yet, but I feel like that was at least important at some point...
> Edit: Oh yeah, and why does `PadNull` make X11 calls? For now I just ifdef'd them out, but I feel like a null api shouldn't need to do anything...

Haha awesome work, but I'm not any kind of developer.. I'm just following this thread to see if there are any test builds available for when something significant happens! Keep up the good work! The last time we had a macOS native PCSX2 was when it was at like version 0.9.5 or earlier. That's why I'm so excited! Cheers!

**Comment by arcum42** [comment]
The main reason pcsx2 uses wxwidgets 3.0 is because that's the version shipped with most Linux distributions. This is largely because the 3.1 branch is technically the unstable branch, though it's supposed to be fairly stable according to the wx guys...

Technically speaking, the GSdx dialog and any other plugin dialogs remaining that aren't wx should probably be ported to wx. We just keep talking about how we'd like to rip out wx, and it'd probably fall on me to port them anyways if we were going to do that.

It's not really vital that padNull works, and it's largely there as a reference. It's based on onepad, though, and on linux, x11 is used for getting the keyboard state. (If you look at keyboard.cpp in onepad, you can see that a lot of it's x11, and the Windows code in there doesn't even work! That's probably the biggest part you'll need to port. I'd be tempted to disable the keyboard portion initially, and just get the controllers working...)

Not absolutely sure on why that one line in GSdx was there...

**Comment by tellowkrinkle** [comment]
@shiggitay 32-bit testing build is now available https://forums.pcsx2.net/Thread-Native-Mac-Testing-Build-Mojave-and-lower

**Comment by shiggitay** [comment]
> @shiggitay 32-bit testing build is now available https://forums.pcsx2.net/Thread-Native-Mac-Testing-Build-Mojave-and-lower

Awesome! I'll slot in my AMD Radeon RX 480 GPU to test it out tomorrow when it's not midnight here.

**Comment by tellowkrinkle** [comment]
Also for anyone trying to build 32-bit applications on macOS, use MacPorts (not Homebrew) to install your packages.  If you add the `+universal` flag when installing, it'll install dylibs with both 32-bit and 64-bit code so you can link against them from both 32-bit and 64-bit applications.

**Comment by shiggitay** [comment]
> Also for anyone trying to build 32-bit applications on macOS, use MacPorts (not Homebrew) to install your packages. If you add the `+universal` flag when installing, it'll install dylibs with both 32-bit and 64-bit code so you can link against them from both 32-bit and 64-bit applications.

Is it possible to compile PCSX2 and its deps etc in 64bit? Slowly more and more people are gonna have Catalina and nothing lower installed on their systems.

**Comment by theLMGN** [comment]
@tellowkrinkle nice work! i just tried it out, and I managed to get the thing to change to SW renderer (after many error messages), now it's complaining it can't find GSdx anymore, and trying to configure it crashes PCSX2

**Comment by tellowkrinkle** [comment]
> I managed to get the thing to change to SW renderer (after many error messages), now it's complaining it can't find GSdx anymore, and trying to configure it crashes PCSX2

Ahh yeah forgot to mention that issue.  Does it work if you restart PCSX2 between GSdx changes?  BTW if you can, try to post issues specific to that build on the forum post instead of here

**Comment by zedr0n** [comment]
as the author of the original port, the recompilers are where the pain is. OSX requires the recompiled data to be aligned which means that I had to manually patch multiple places in nasm and things still broke on a regular basis. Maybe they finally fixed it in later versions of OSX

**Comment by tellowkrinkle** [comment]
The issue is 16 byte stack alignment when calling external functions right?  So far nothing has blown up so maybe things haven't changed too much since the original port?  Admittedly it's not too common for the recompilers to call out into external functions...
As for the specific GCC version, osx no longer uses GCC and everything's been compiling fine on their recent versions of clang

**Comment by lonkle** [comment]
It's been a year since this issue has been actively discussed but it looks like there was some good progress. Obviously software rendering is all that can be done without Vulkan, but I hear PCSX2 is not against using Vulkan so one day...a full-speed macOS PCSX2. I honestly never thought it would happen, but I'm pretty happy.

**Comment by MaddTheSane** [comment]
A Metal renderer would be better, and yes, @tellowkrinkle has made initial work on the Metal back-end. That and Vulkan support has been added to the master branch.

As for arm64 support: That would require a lot of work, as there are many places where PCSX2 uses Intel intrinsics (I think the software renderer uses them, for instance). Also, an arm64 interpreter would need to be created. I think someone is working on an arm64 port for Linux, which should help with an M1 port.

**Comment by Spidy123222** [comment]
> A Metal renderer would be better, and yes, @tellowkrinkle has made initial work on the Metal back-end. That and Vulkan support has been added to the master branch.
> 
> 
> 
> As for arm64 support: That would require a lot of work, as there are many places where PCSX2 uses Intel intrinsics (I think the software renderer uses them, for instance). Also, an arm64 interpreter would need to be created. I think someone is working on an arm64 port for Linux, which should help with an M1 port.

A metal backend is always likely but using moltenvk should be enough since Kronos constantly updates it o make Vulkan viable on such a platform.

**Comment by seta-san** [comment]
@tellowkrinkle how "ported" to macos x would you say pcsx2 is now ?

**Comment by tellowkrinkle** [comment]
Remaining issues:
- Software renderer needs to default to not-OpenGL
  - We should probably also remove the OpenGL option completely, since it's not going to work
- Keyboard shortcuts, fixed by #5792
- CI currently builds with packages downloaded from Homebrew, which aren't compatible with macOS versions older than the version the CI is running.  We need to build our own dependencies with the flags set to build with support for our target  oldest macOS version (currently 10.13)
- CI needs to package MoltenVK into the app bundle

**Comment by seta-san** [comment]
That was fast. https://github.com/PCSX2/pcsx2/pull/5807/files

**Comment by tellowkrinkle** [comment]
MacOS builds are live!
